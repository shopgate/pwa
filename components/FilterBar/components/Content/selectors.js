import { createSelector } from 'reselect';
import {
  getCategoryProductCount,
  getCurrentCategoryChildCount,
} from '@shopgate/pwa-common-commerce/category/selectors';
import { getProductsResult } from '@shopgate/pwa-common-commerce/product/selectors/product';
import { hasActiveFilters } from '@shopgate/pwa-common-commerce/filter/selectors';

/**
  * Checks if the filter bar is shown within a category page
  * @return {bool}
  */
export const isFilterBarShown = createSelector(
  getCurrentCategoryChildCount,
  getCategoryProductCount,
  hasActiveFilters,
  getProductsResult,
  (subCategories, categoryProductCount, activeFilters, { totalProductCount }) => {
    const hasProducts = (totalProductCount === null && categoryProductCount > 0) ||
       totalProductCount > 0;

    return !subCategories && (hasProducts || activeFilters);
  }
);
