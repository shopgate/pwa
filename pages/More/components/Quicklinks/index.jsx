import React from 'react';
import PropTypes from 'prop-types';
import Section from '../Section';
import connect from './connector';

/**
 * The Quicklinks component.
 * @param {Object} props The component props.
 * @param {Array} props.entries The quicklinks.
 * @returns {JSX}
 */
function Quicklinks({ entries }) {
  if (!entries || !entries.length) {
    return null;
  }

  return (
    <Section title="navigation.more_menu">
      {entries.map(entry => (
        <Section.Item href={entry.url} key={entry.url} label={entry.label} />
      ))}
    </Section>
  );
}

Quicklinks.propTypes = {
  entries: PropTypes.arrayOf(PropTypes.shape()),
};

Quicklinks.defaultProps = {
  entries: null,
};

export default connect(Quicklinks);
