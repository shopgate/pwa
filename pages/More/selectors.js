import { createSelector } from 'reselect';

/**
 * Selects the menu state from the store.
 * @param {Object} state The global state.
 * @return {Object}
 */
const getMenuState = state => state.menu.menusById;

/**
 * @returns {Function}
 */
export const makeGetMenuById = () => (
  createSelector(
    getMenuState,
    (state, props) => props.id,
    (menus, id) => {
      if (!id || !menus[id] || !menus[id].entries) {
        return null;
      }

      return menus[id].entries;
    }
  )
);
