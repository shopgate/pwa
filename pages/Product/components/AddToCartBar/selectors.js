import { createSelector } from 'reselect';

/**
 * Selects the addToCartBar state.
 * @param {Object} state The current application state.
 * @return {Object}
 */
const getAddToCartBarState = state => state.ui.addToCartBar;

/**
 * Selects the current added count from the addToCartBar state.
 * @param {Object} state The current application state.
 * @return {number}
 */
export const selectActionCount = createSelector(
  getAddToCartBarState,
  state => state.added
);

/**
 * Returns whether the addToCartBar is visible or not.
 * @param {Object} state The application state.
 * @returns {boolean}
 */
export const isVisible = createSelector(
  getAddToCartBarState,
  state => state.visible
);
