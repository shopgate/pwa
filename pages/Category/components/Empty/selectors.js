import { createSelector } from 'reselect';
import { isNumber } from '@shopgate/pwa-common/helpers/validation';
import {
  getCategoryProductCount,
  getCurrentCategoryChildCount,
} from '@shopgate/pwa-common-commerce/category/selectors';
import { getProductsResult } from '@shopgate/pwa-common-commerce/product/selectors/product';

/**
 * Determines whether or not a category has any content,
 * (subcategories, child products or products received via request).
 */
export const isCategoryEmpty = createSelector(
  getCategoryProductCount,
  getCurrentCategoryChildCount,
  getProductsResult,
  (categoryProductCount, categoryChildrenCount, { products, totalProductCount }) => {
    /**
     * Check if we have any products and if we have a valid number
     * for the total number of products. Not having a valid number
     * indicates that a request is ongoing.
     */
    const hasProductResult = products.length === 0 && isNumber(totalProductCount);

    /**
     * Check if we have any subcategories, child products or a product result for this category.
     */
    if (categoryChildrenCount === 0 && (categoryProductCount === 0 || hasProductResult)) {
      return true;
    }

    return false;
  }
);
