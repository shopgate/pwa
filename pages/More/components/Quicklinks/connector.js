import { connect } from 'react-redux';
import { QUICKLINKS_MENU } from '@shopgate/pwa-common/constants/MenuIDs';
import { getMenuById } from '@shopgate/pwa-common/selectors/menu';

const props = {
  id: QUICKLINKS_MENU,
};

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  entries: getMenuById(state, props),
});

export default connect(mapStateToProps);
