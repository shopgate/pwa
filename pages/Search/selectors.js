import { createSelector } from 'reselect';
import { hasActiveFilters } from '@shopgate/pwa-common-commerce/filter/selectors';
import { getResultByHash } from '@shopgate/pwa-common-commerce/product/selectors/product';

export const showNoResults = createSelector(
  getResultByHash,
  (results) => {
    if (results === null) {
      return false;
    }

    if (results && results.isFetching) {
      return false;
    }

    if (results && !results.isFetching && (results.products && results.products.length > 0)) {
      return false;
    }

    return true;
  }
);

export const areResultsFetching = createSelector(
  getResultByHash,
  (results) => {
    if (!results || results.isFetching) {
      return true;
    }

    return false;
  }
);

export const showFilterBar = createSelector(
  hasActiveFilters,
  areResultsFetching,
  showNoResults,
  (hasFilters, isFetching, hasResults) => {
    if (isFetching) {
      return false;
    }

    if (!hasFilters && hasResults) {
      return false;
    }

    return true;
  }
);
