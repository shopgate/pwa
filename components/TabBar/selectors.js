import { createSelector } from 'reselect';
import { getHistoryPathname } from '@shopgate/pwa-common/selectors/history';
import { INDEX_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import { CART_PATH } from '@shopgate/pwa-common-commerce/cart/constants';
import { FAVORITES_PATH } from '@shopgate/pwa-common-commerce/favorites/constants';
import { CATEGORY_PATH } from '@shopgate/pwa-common-commerce/category/constants';
import { SEARCH_PATH } from '@shopgate/pwa-common-commerce/search/constants';
import { BROWSE_PATH } from 'Pages/Browse/constants';
import { MORE_PATH } from 'Pages/More/constants';
import {
  TAB_HOME,
  TAB_BROWSE,
  TAB_CART,
  TAB_FAVORITES,
  TAB_MORE,
  TAB_NONE,
} from './constants';

/**
 * Returns a tabBar state.
 * @param {Object} state State.
 * @return {Object}
 */
const getTabBarState = state => state.ui.tabBar;

/**
 * Returns what tab is active, returns TAB_NONE if none is active.
 * @param {Object} state The application state.
 * @returns {string}
 */
export const getActiveTab = createSelector(
  getHistoryPathname,
  (pathname) => {
    switch (true) {
      case pathname === INDEX_PATH:
        return TAB_HOME;
      case (pathname === BROWSE_PATH
        || pathname.startsWith(SEARCH_PATH)
        || pathname.startsWith(CATEGORY_PATH)):
        return TAB_BROWSE;
      case pathname === CART_PATH:
        return TAB_CART;
      case pathname === MORE_PATH:
        return TAB_MORE;
      case pathname === FAVORITES_PATH:
        return TAB_FAVORITES;
      default:
        return TAB_NONE;
    }
  }
);

/**
 * Checks if the tab bar is currently enabled.
 * @return {boolean}
 */
export const isTabBarEnabled = createSelector(
  getTabBarState,
  state => state.enabled
);

/**
 * Checks if the tab bar is currently visible.
 * @return {boolean}
 */
export const isTabBarVisible = createSelector(
  getTabBarState,
  isTabBarEnabled,
  (state, enabled) => {
    if (!enabled) {
      return false;
    }

    return state.visible;
  }
);
