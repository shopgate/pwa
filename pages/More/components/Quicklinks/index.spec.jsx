import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { createMockStore } from '@shopgate/pwa-common/store';
import { getMenuById } from '@shopgate/pwa-common/selectors/menu';
import Quicklinks from './index';

const store = createMockStore();

jest.mock('@shopgate/pwa-common/components/Link', () => {
  /* eslint-disable react/prop-types, require-jsdoc */
  const Link = ({ children }) => (
    <div>
      {children}
    </div>
  );

  /* eslint-enable react/prop-types, require-jsdoc */
  return Link;
});

let mockedQuicklinks;
jest.mock('@shopgate/pwa-common/selectors/menu', () => ({
  getMenuById: () => mockedQuicklinks,
}));

describe('<Quicklinks />', () => {
  beforeEach(() => {
    mockedQuicklinks = [
      /* eslint-disable extra-rules/no-single-line-objects */
      { url: '/some/url', label: 'Some Label' },
      { url: '/another/url', label: 'Another Label' },
      /* eslint-enable extra-rules/no-single-line-objects */
    ];
  });

  it('should render quicklinks', () => {
    const quicklinks = getMenuById();
    const wrapper = mount((
      <Provider store={store}>
        <Quicklinks />
      </Provider>));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Quicklinks').isEmptyRender()).toBe(false);
    expect(wrapper.find('Headline').text()).toBe('navigation.more_menu');

    const links = wrapper.find('Link');
    quicklinks.forEach((entry, index) => {
      const link = links.at(index);
      expect(link.text()).toBe(entry.label);
      expect(link.prop('href')).toBe(entry.url);
    });
  });

  it('should not render when no quicklinks are there', () => {
    mockedQuicklinks = [];
    const wrapper = mount((
      <Provider store={store}>
        <Quicklinks />
      </Provider>));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Quicklinks').isEmptyRender()).toBe(true);
  });
});
