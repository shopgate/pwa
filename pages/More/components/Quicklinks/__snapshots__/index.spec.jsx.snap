// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Quicklinks /> should not render when no quicklinks are there 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Quicklinks)>
    <Quicklinks
      dispatch={[Function]}
      entries={Array []}
    />
  </Connect(Quicklinks)>
</Provider>
`;

exports[`<Quicklinks /> should render quicklinks 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(Quicklinks)>
    <Quicklinks
      dispatch={[Function]}
      entries={
        Array [
          Object {
            "label": "Some Label",
            "url": "/some/url",
          },
          Object {
            "label": "Another Label",
            "url": "/another/url",
          },
        ]
      }
    >
      <Section
        title="navigation.more_menu"
      >
        <Headline
          style={
            Object {
              "margin": "24px 20px 16px",
            }
          }
          tag="h2"
          text="navigation.more_menu"
        >
          <h2
            className="css-18cwfw7"
            data-test-id="Headline"
            style={
              Object {
                "margin": "24px 20px 16px",
              }
            }
          >
            <Translate
              className=""
              params={Object {}}
              string="navigation.more_menu"
            >
              <span
                className=""
              >
                navigation.more_menu
              </span>
            </Translate>
          </h2>
        </Headline>
        <div
          className={
            Object {
              "data-css-13ch7sr": "",
            }
          }
        >
          <MoreMenuItem
            className={null}
            href="/some/url"
            key="/some/url"
            label="Some Label"
            onClick={null}
            testId={null}
          >
            <Link
              className="css-1hj8a57"
              href="/some/url"
            >
              <div>
                <Translate
                  className=""
                  params={Object {}}
                  string="Some Label"
                >
                  <span
                    className=""
                  >
                    Some Label
                  </span>
                </Translate>
              </div>
            </Link>
          </MoreMenuItem>
          <MoreMenuItem
            className={null}
            href="/another/url"
            key="/another/url"
            label="Another Label"
            onClick={null}
            testId={null}
          >
            <Link
              className="css-1hj8a57"
              href="/another/url"
            >
              <div>
                <Translate
                  className=""
                  params={Object {}}
                  string="Another Label"
                >
                  <span
                    className=""
                  >
                    Another Label
                  </span>
                </Translate>
              </div>
            </Link>
          </MoreMenuItem>
        </div>
      </Section>
    </Quicklinks>
  </Connect(Quicklinks)>
</Provider>
`;
